### PR TITLE
Refine dispatcher and block sync

### PR DESCRIPTION
**Pull request by CoderZhi**
refine the block sync related code

### CODE REVIEW

File: blocksync/blocksync.go [hunk 1]
```diff
@@ -196,7 +196,7 @@ func (bs *blockSyncer) ProcessSyncRequest(ctx context.Context, peer peerstore.Pe
 		syncCtx, cancel := context.WithTimeout(ctx, bs.processSyncRequestTTL)
 		defer cancel()
 		if err := bs.unicastHandler(syncCtx, peer, blk.ConvertToBlockPb()); err != nil {
-			log.L().Debug("Failed to response to ProcessSyncRequest.", zap.Error(err))
```
Comment (raullenchai) [inline blocksync/blocksync.go:208]: seem useful to keep this one.

File: dispatcher/dispatcher.go [hunk 2]
```diff
@@ -348,20 +371,12 @@ func (d *IotxDispatcher) HandleTell(ctx context.Context, chainID uint32, peer pe
...
 
-func (d *IotxDispatcher) enqueueEvent(event interface{}) {
-	if len(d.eventChan) == cap(d.eventChan) {
-		log.L().Warn("dispatcher event chan is full, drop an event.")
-		return
-	}
-	d.eventChan <- event
-}
-
 func (d *IotxDispatcher) updateEventAudit(t iotexrpc.MessageType) {
```
Comment (raullenchai) [inline dispatcher/dispatcher.go:380]: should update this stat as well, can defer to the next pr

File: blocksync/blocksync.go [hunk 3]
```diff
@@ -152,32 +150,20 @@ func (bs *blockSyncer) Stop(ctx context.Context) error {
 
 // ProcessBlock processes an incoming latest committed block
```
Comment (raullenchai) [inline blocksync/blocksync.go:151]: Update comment

**Review by raullenchai — APPROVED** [review]

File: blocksync/blocksync.go (hunk 3 above)
Comment (CoderZhi) [reply blocksync/blocksync.go]: addressed

File: dispatcher/dispatcher.go (hunk 2 above)
Comment (CoderZhi) [reply dispatcher/dispatcher.go:382]: addressed

File: dispatcher/dispatcher.go [hunk 4]
```diff
@@ -255,15 +280,16 @@ func (d *IotxDispatcher) handleBlockSyncMsg(m *blockSyncMsg) {
...
-	d.subscribersMU.RUnlock()
-	if ok {
+	if subscriber := d.subscriber(m.ChainID()); subscriber != nil {
 		d.updateEventAudit(iotexrpc.MessageType_BLOCK_REQUEST)
 		// dispatch to block sync
 		if err := subscriber.HandleSyncRequest(m.ctx, m.peer, m.sync); err != nil {
 			log.L().Error("Failed to handle sync request.", zap.Error(err))
 		}
+		d.syncChanLock.RLock()
+		defer d.syncChanLock.RUnlock()
```
Comment (dustinxie) [inline dispatcher/dispatcher.go:290]: this is not necessary?
only `HandleTell()` calls `handleBlockSyncMsg`

File: dispatcher/dispatcher.go [hunk 5]
```diff
@@ -274,41 +300,81 @@ func (d *IotxDispatcher) dispatchAction(ctx context.Context, chainID uint32, msg
...
-		chainID: chainID,
-		action:  (msg).(*iotextypes.Action),
-	})
+	subscriber := d.subscriber(chainID)
+	if subscriber == nil {
+		log.L().Debug("no subscriber for this chain id, drop the action", zap.Uint32("chain id", chainID))
+		return
+	}
+	d.actionChanLock.Lock()
+	defer d.actionChanLock.Unlock()
```
Comment (dustinxie) [inline dispatcher/dispatcher.go:309]: this is not necessary?
only `HandlBroadcast()` calls `dispatchAction` that will send to this channel

File: dispatcher/dispatcher.go [hunk 6]
```diff
@@ -274,41 +300,81 @@ func (d *IotxDispatcher) dispatchAction(ctx context.Context, chainID uint32, msg
...
-		chainID: chainID,
-		block:   (msg).(*iotextypes.Block),
-	})
+	subscriber := d.subscriber(chainID)
+	if subscriber == nil {
+		log.L().Debug("no subscriber for this chain id, drop the block", zap.Uint32("chain id", chainID))
+		return
+	}
+	d.blockChanLock.Lock()
+	defer d.blockChanLock.Unlock()
```
Comment (dustinxie) [inline dispatcher/dispatcher.go:336]: this is needed,
both `HandleBroadcast()` and `HandleTell()` call `dispatchBlockCommit` that will send to this channel

**Pull request merged**